torch_xla/csrc/pooling.cpp

```diff
@@ -538,59 +540,92 @@ xla::XlaOp BuildAvgPoolNdBackward(xla::XlaOp out_backprop, xla::XlaOp input,
+  const auto input_size = XlaHelpers::SizesOfXlaOp(input);
+  XLA_CHECK(input_size.size() == 4 || input_size.size() == 3)
+      << "Only 4D or 3D tensors supported";
+  return BuildAdaptiveAvgPool(input, output_size, /*pool_dim=*/2);
+}
+
+xla::XlaOp BuildAdaptiveAvgPoolBackward(xla::XlaOp out_backprop,
+                                        xla::XlaOp input,
+                                        int unsigned pool_dim) {
   BatchInput batch_out_backprop_info =
-      CreateBatchInput(/*input=*/out_backprop, /*spatial_dim_count=*/2);
+      CreateBatchInput(/*input=*/out_backprop, /*spatial_dim_count=*/pool_dim);
   const auto out_backprop_size =
       XlaHelpers::SizesOfXlaOp(batch_out_backprop_info.batch_input);
-  XLA_CHECK_EQ(out_backprop_size.size(), 4)
+  XLA_CHECK_EQ(out_backprop_size.size(), pool_dim + 2)
       << "Invalid rank of gradient output";
-  std::vector<xla::int64> output_size{out_backprop_size[2],
-                                      out_backprop_size[3]};
+  std::vector<xla::int64> output_size(out_backprop_size.begin() + 2,
+                                      out_backprop_size.end());
   auto gradients_size = XlaHelpers::SizesOfXlaOp(input);
-  XLA_CHECK(gradients_size.size() == 4 || gradients_size.size() == 3)
-      << "Only 4D or 3D tensors supported";
-  if (gradients_size.size() == 3) {
+  if (gradients_size.size() == pool_dim + 1) {
     gradients_size.insert(gradients_size.begin(), 1);
   }
   const auto kernel_size =
-      AdaptiveAvgPoolKernelSize(gradients_size, output_size);
-  std::vector<std::pair<xla::int64, xla::int64>> no_padding(2);
+      AdaptiveAvgPoolKernelSize(gradients_size, output_size, pool_dim);
+  std::vector<std::pair<xla::int64, xla::int64>> no_padding(pool_dim);
   xla::XlaOp batch_result = xla::AvgPoolGrad(
       /*out_backprop=*/batch_out_backprop_info.batch_input,
       /*gradients_size=*/gradients_size,
       /*kernel_size=*/kernel_size,
       /*stride=*/kernel_size,
       /*spatial_padding=*/no_padding,
-      /*data_format=*/MakeNCHWFormat(2),
+      /*data_format=*/MakeNCHWFormat(pool_dim),
       /*counts_include_padding=*/false);
   return RemoveTrivialBatch(
       /*batch=*/batch_result,
       /*original_rank=*/batch_out_backprop_info.original_rank,
-      /*spatial_dim_count=*/2);
+      /*spatial_dim_count=*/pool_dim);
+}
+
+xla::XlaOp BuildAdaptiveAvgPool3dBackward(xla::XlaOp out_backprop,
+                                          xla::XlaOp input) {
+  auto gradients_size = XlaHelpers::SizesOfXlaOp(input);
+  XLA_CHECK(gradients_size.size() == 4 || gradients_size.size() == 5)
+      << "Only 4D or 5D tensors supported";
+  return BuildAdaptiveAvgPoolBackward(out_backprop, input,
+                                      /*pool_dim=*/3);
+}
+
+xla::XlaOp BuildAdaptiveAvgPool2dBackward(xla::XlaOp out_backprop,
+                                          xla::XlaOp input) {
+  auto gradients_size = XlaHelpers::SizesOfXlaOp(input);
```

Comment:
Same consideration WRT `XlaHelpers::ShapeOfXlaOp()`.

Reply:
I think gradient_size vector is also passed as an input to the  xla::AvgPoolGrad. And I can't change the interface of that function. We are padding 1 to the gradience_size vector in case the gradient_input has no batch dimension. This means we need create a local vector, insert one dimension if needed and pass to the xla::AvgPoolGrad.  

If I use ShapeOfXlaOp I need use .dimensions method as you said earlier and create that local vector copy to perform the insertion.
If I change the interface of AdaptiveAvgPoolKernelSize to take xla::Shape& as input instead of the vector. Here is what the code look like:

xla::Shape& gradient_shape = gradients_size = XlaHelpers::ShapeOfXlaOp(input)
const auto gradient_size = gradien_shape.dimensions()
// Do the insert one dimension if no batch dimension
// Pass gradient_shape to AdaptiveAvgPoolKernelSize ???? [ Because if I implement the previous comment this function is accepting shape, but the problem is that dimensions in this shape can still be missing the batch dimension => check the dimension in the AdaptiveAvgPoolKernelSize function and insert there as well???
// Pass the gradient_size vector to xla::AvgPoolGrad [ Can not pass gradient_shape.dimensions()

I see using shape reference is introducing more sub-optimalities thank making it optimal.
What am I missing?

Reply:
No. In this function, as well as many others you added, a vector is not changed, and hence it can use the shape.
Clearly if the vector needs to be changed you need a vector.
There are a few other places which need to be changed in that file, but your PR should not make things worse.
The `AdaptiveAvgPoolKernelSize()` for example, already takes an `absl::Span`, which is what `xla::Shape::dimensions()` returns.

Reply:
https://github.com/pytorch/xla/blob/master/torch_xla/csrc/pooling.cpp#L576

Reply:
This is the line in 2d backward implementation as is now. I have kept that same in the generalized implementation. This is where the vector was being changed, and is being changed in the PR.

Reply:
This is different. This API (the one this comment is attached to) does not change the dimensions vector.

Reply:
I think I am missing something. I can push the commit with this understanding and you can let me know if that's what you were thinking. I can always undo, correct it further.

Reply:
This is still not fixed WRT vector-creation vs. const-shape

Reply:
fixed now.